### DISCUSSION TITLE
Qual a melhor maneira de usar componentes no React?

### DISCUSSION THREAD

**Discussion by wilsonneto-swe** [frontendbr/forum#1559] [Discussão]
Galera, usar componentes em react com function ou com class? Qual a maneira correta? Ou ambas são corretas?

Fazendo a semana omnistack esta semana percebi a preferencia deles por usar componentes funcionais, todos declarados em forma de função.

Como eu venho do C#, lido bem com classes e acho até mais organizados e mais fácil de ler.

Há uma maneira considerada a correta? 

Obrigado!

**Comment by bizarro** [comment]
A resposta rápida é que ambas são corretas. 😄 Mas eu, particularmente, adoro usar eles como classes pois me lembram as ES6 classes, e em projetos sem React eu costumo usar bastante.

Mas a tendência que ando observando é o replacement de componentes que usam classes pelos funcionais em projetos React mesmo. Na maioria dos casos fica bem menos verbosos e você escreve bem menos.

Então minha sugestão é estudar para saber fazer das duas maneiras. E prestar bastante atenção na API de `useState`, `useEffect`, `useCallback`, etc. Pois ficam bem diferentes na implementação do que um componente que utiliza extend de `PureComponent` ou `Component`.

**Comment by tcelestino** [comment]
@wilsonneto-dev, se preferir, pode ler esse [artigo](https://overreacted.io/pt-br/how-are-function-components-different-from-classes/) do Dan Abramov sobre o desenvolvimento de componentes em classes e função. É bem completo e vai conseguir os beneficios de cada uma das implementações.

**Comment by davipm** [comment]
Ambas arbodagens são corretas, mas os Hooks facilitam demais a vida, e são menos verbosos e bem fácil a reutilização da lógica com os Hooks custom, eu já não uso mais class nos meu projetos novos.

**Comment by wilsonneto-swe** [comment]
Opa, @davi-94, @tcelestino, @lhbzr , muito obrigado pelas respostas!

**Comment by victorferreira** [comment]
Um ponto que eu considero importante a ser considerado sobre classes é que no JS o modelo de objetos é baseado em _proptotype_, o que é diferente de várias outras linguagens orientadas a objetos. O uso de _class_ no javascript é apenas _syntax sugar_ para o _prototype_, então vale a pena levar isso em consideração. 

Sobre os componentes do react, eu prefiro usar _functional components_ já que casa bem da ideia do React de ter componentes que reagem em função da mudança do estado. Nesse sentido não existe certo ou errado, mas o que faz sentido pra você e para o seu cenário. Como diria o filósofo, cada escolha uma renúncia, isso é a vida.

**Comment by ninetails** [comment]
Também sou fã de _componentes funcionais_, porém concordo que ambas as formas são corretas.

Adicionando, diria para levar em conta o perfil do time/seu perfil. Como nada impede de utilizar ambas as formas, tente utilizar a que vc se sinta mais habituado. Use classes se achar que fica mais à vontade com orientação a objeto ou componentes funcionais se caso quiser se desafiar e tentar algo mais "simplificado" (menos verboso) ou curta programação funcional.

Num lugar que trabalhei anteriormente tinha um time que utilizava classes e TypeScript pois tinha muito Javeiro lá.

**Comment by felipefialho** [comment]
Alterei o titulo e a dúvida original para dar mais margem de discussão na issue (que curti bastante) 😄
